### PR TITLE
RavenDB-14555 Moving to per core behavior

### PR DESCRIPTION
**Pull request by ayende**

### CODE REVIEW

File: src/Sparrow/Json/JsonContextPoolBase.cs [hunk 1]
```diff
@@ -388,32 +152,90 @@ public void Dispose()
...
+                if (Interlocked.CompareExchange(ref core[i], context, null) == null)
                     return;
             }
+
+            if (core.Length < 64)
+            {
+                var newCore = new T[core.Length * 2];
+                Array.Copy(core, 0, newCore, 0, core.Length);
+                newCore[core.Length] = context;
+                if (Interlocked.CompareExchange(ref _perCoreContexts[currentProcessorId], newCore, core) == core)
```
Comment (karmeli87) [inline src/Sparrow/Json/JsonContextPoolBase.cs]: I think you might leak contexts here. 
If someone put the context back, after you copied it, but before the exchange.

File: src/Sparrow/Json/JsonContextPoolBase.cs [hunk 2]
```diff
@@ -426,27 +248,49 @@ public virtual void Dispose()
...
                 _disposed = true;
-                CleanThreadLocalState();
-                ThreadLocalCleanup.ReleaseThreadLocalState -= CleanThreadLocalState;
-                _nativeMemoryCleaner.Dispose();
-                _nativeMemoryCleaner = null;
+                _idleTimer.Dispose();
+
+                while (_globalQueue.TryDequeue(out var result))
+                {
+                    result.Dispose();
```
Comment (karmeli87) [inline src/Sparrow/Json/JsonContextPoolBase.cs:246]: Do we need to check if in use?

File: src/Sparrow/Json/JsonContextPoolBase.cs (hunk 1 above)
Comment (ayende) [reply src/Sparrow/Json/JsonContextPoolBase.cs]: Good point, I think this is going to be really rare thing, and will be covered by the finalizer. But will think on how to resolve this.

File: src/Sparrow/Json/JsonContextPoolBase.cs (hunk 2 above)
Comment (ayende) [reply src/Sparrow/Json/JsonContextPoolBase.cs:246]: This is in the `Dispose`, caller is responsible for ensuring this.

**Comment by ayende** [comment]
Looks good to me

**Pull request merged**